src/NumericalAlgorithms/Interpolation/MultiLinearSpanInterpolation.hpp

```diff
@@ -0,0 +1,397 @@
+  // Use linear extrapolation based of the lowest
+  // two points in the table
+  ASSERT(allow_extrapolation_below_data[whichDimension] or
+             UNLIKELY(relativeCoordinate > 0.),
+         "Interpolation exceeds lower table bounds.");
+
+  // We are exceeding the table bounds:
+  // Use linear extrapolation based of the highest
+  // two points in the table
+
+  ASSERT(allow_extrapolation_above_data[whichDimension] or
+             UNLIKELY(currentIndex + 1 < numberOfPoints[whichDimension]),
+         "Interpolation exceeds upper table bounds.");
+
+  // Enforce index ranges
+  currentIndex =
+      std::min(numberOfPoints[whichDimension] - 2, std::max(0ul, currentIndex));
+
+  return currentIndex;
+}
+
+/// Compute interpolation weights for 1D tables
+template <size_t Dimension, size_t NumberOfVariables, bool UniformSpacing>
+typename MultiLinearSpanInterpolation<Dimension, NumberOfVariables,
+                             UniformSpacing>::template weight_t<1>
+MultiLinearSpanInterpolation<Dimension, NumberOfVariables,
+                             UniformSpacing>::getWeights(const double x1)
+    const {
+  double xx;
+
+  weight_t<1> weights;
+
+  auto index = Index<1>(findIndex(0, x1));
+
+  if constexpr (UniformSpacing) {
+    xx = (x1 - x[0][index[0]]) * inverseSpacing[0];
+  } else {
+    xx = (x1 - x[0][index[0]]) / (x[0][index[0] + 1] - x[0][index[0]]);
+  }
+
+  // Note: first index varies fastest
+  weights.weights = std::array<double, 2>({
+      (1. - xx),  // 0
+      xx,         // 1
+  });
+
+  // Compute indices
+
+  weights.index[0] = index[0];
+  weights.index[1] = index[0] + 1;
+
+  return weights;
+}
+
+/// Compute interpolation weights for 2D tables
+
+template <size_t Dimension, size_t NumberOfVariables, bool UniformSpacing>
+typename MultiLinearSpanInterpolation<Dimension, NumberOfVariables,
+                             UniformSpacing>::template weight_t<2>
+MultiLinearSpanInterpolation<Dimension, NumberOfVariables,
+                             UniformSpacing>::getWeights(const double x1,
+                                                         const double x2)
+    const {
+  weight_t<2> weights;
+
+  auto index = Index<2>(findIndex(0, x1), findIndex(1, x2));
+
+  double xx, yy;
+  if constexpr (UniformSpacing) {
+    xx = (x1 - x[0][index[0]]) * inverseSpacing[0];
+    yy = (x2 - x[1][index[1]]) * inverseSpacing[1];
+  } else {
+    xx = (x1 - x[0][index[0]]) / (x[0][index[0] + 1] - x[0][index[0]]);
+    yy = (x2 - x[1][index[1]]) / (x[1][index[1] + 1] - x[1][index[1]]);
+  }
+
+  // Note: first index varies fastest
+  weights.weights = std::array<double, 4>({
+      (1. - xx) * (1. - yy),  // 00
+      xx * (1. - yy),         // 10
+      (1. - xx) * yy,         // 01
+      xx * yy,                // 11
+  });
+
+  // Compute indices
+  //
+  for (size_t j = 0; j < 2; ++j)
+    for (size_t i = 0; i < 2; ++i) {
+      auto tmp_index = Index<2>(index[0] + i, index[1] + j);
```

Comment:
why not `const Index<2> tmp_index(index[0] + i, index[1] + j);`?

Reply:
I'm an "always auto" person :)